### PR TITLE
Issue #999 : Adds Percentage progress bar for attachment downloads

### PR DESCRIPTION
**Pull request by harjot-oberai**
#This PR implements a percentage progress dialog that is displayed when an attachment is being downloaded.

GIF

### CODE REVIEW

File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java [hunk 1]
```diff
@@ -32,6 +34,8 @@
     private static final String HEADER_PARAM_FORMAT = "format";
     private static final String HEADER_FORMAT_FLOWED = "flowed";
 
+    private static int filledBytes = 0;
```
Comment (philipwhiuk) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java]: A Utility class should not store temporal information. This needs some more thought.

Comment (harjot-oberai) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java]: I thought of another approach, it is as follows : 
1. Create a `ProgressCallback` interface with one method `onUpdate()`.
2. Create a copy of `createBody()` function that takes an object of `ProgressCallback` as an extra parameter and calls `onUpdate()` every 50 millisecond while the input stream is being copied.
3. In the implementation of the `onUpdate` method we can update the progress in the dialog.

Is this approach correct ?

Comment (philipwhiuk) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java]: Seems like a better solution to me.

Comment (harjot-oberai) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java]: Cool 👍

**Comment by harjot-oberai** [comment]
@philipwhiuk : I have implemented the requested changes and pushed them yesterday, please review them and suggest any changes that might be necessary.

**Comment by Valodim** [comment]
Did you see MessagingController already has a callback mechanism? The dialog fragment could simply register itself as a listener (`MessagingController.getInstance(context).addListener(this)`) and get progress updates that way.

**Comment by harjot-oberai** [comment]
Sorry, but I don't get what you are trying to say. Are you saying that the implementation can be done, without using `AttachmentProgressCallback` ?

**Comment by Valodim** [comment]
Not sure about that, but it should work without `AttachmentDownloadDialogFragment.updateProgress`

**Comment by harjot-oberai** [comment]
@Valodim : I have implemented the call back using MessagingController. Please review and suggest any other changes

**Comment by harjot-oberai** [comment]
@Valodim @philipwhiuk : Did anyone review this ?

File: k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java [hunk 2]
```diff
@@ -860,6 +861,85 @@ public void fetchPart(Message message, Part part, MessageRetrievalListener<Messa
         }
     }
 
+    @Override
+    public void fetchPartWithProgressCallback(Message message, Part part, MessageRetrievalListener<Message> listener, AttachmentProgressCallback progressCallback) throws MessagingException {
```
Comment (philipwhiuk) [inline k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java]: This is a lot of new code in one large function. Is it a slightly altered version of another function? If so, can we reduce the code duplication please.

Comment (harjot-oberai) [inline k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java]: Yes, I think that is possible. I only need to pass the callback, so I can add a check for the `progressCallback` being null in the same code. Will do it shortly.

**Comment by harjot-oberai** [comment]
@philipwhiuk : Implemented the requested changes 👍. Anything else ?

File: k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java [hunk 3]
```diff
@@ -919,8 +840,13 @@ public void fetchPartWithProgressCallback(Message message, Part part, MessageRet
                             String contentTransferEncoding =
                                     part.getHeader(MimeHeader.HEADER_CONTENT_TRANSFER_ENCODING)[0];
                             String contentType = part.getHeader(MimeHeader.HEADER_CONTENT_TYPE)[0];
-                            MimeMessageHelper.setBody(part, MimeUtility.createBodyWithProgressCallback(bodyStream, contentTransferEncoding,
-                                    contentType, progressCallback));
+                            if(progressCallback != null) {
```
Comment (philipwhiuk) [inline k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java]: Is this null check necessary? You're passing either value either way.

File: k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java [hunk 4]
```diff
@@ -839,8 +840,13 @@ public void fetchPart(Message message, Part part, MessageRetrievalListener<Messa
                             String contentTransferEncoding =
                                     part.getHeader(MimeHeader.HEADER_CONTENT_TRANSFER_ENCODING)[0];
                             String contentType = part.getHeader(MimeHeader.HEADER_CONTENT_TYPE)[0];
-                            MimeMessageHelper.setBody(part, MimeUtility.createBody(bodyStream, contentTransferEncoding,
-                                    contentType));
+                            if(progressCallback != null) {
```
Comment (philipwhiuk) [inline k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java]: This check isn't necessary - if it's null you're passing null, if it's not null you're passing the value.

File: k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java (hunk 3 above)
Comment (philipwhiuk) [inline k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java]: Hmm, I see - you're calling createBody twice... I don't really see why.

Comment (harjot-oberai) [inline k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java]: Yes, you're right, there is no need to check the null condition. Removing it now.

**Comment by harjot-oberai** [comment]
@philipwhiuk : done 👍

File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java [hunk 5]
```diff
@@ -990,24 +993,37 @@ public static boolean isDefaultMimeType(String mimeType) {
...
         if (MimeUtil.isMessage(contentType)) {
             tempBody = new BinaryTempFileMessageBody(contentTransferEncoding);
         } else {
             tempBody = new BinaryTempFileBody(contentTransferEncoding);
         }
 
         OutputStream out = tempBody.getOutputStream();
+        Timer timer = null;
         try {
+            if(progressCallback != null){
```
Comment (cketti) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java]: Code style: Missing spaces.

See https://github.com/k9mail/k-9/wiki/CodeStyle

File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java [hunk 6]
```diff
@@ -990,24 +993,37 @@ public static boolean isDefaultMimeType(String mimeType) {
...
+                timer = new Timer();
+                timer.scheduleAtFixedRate(new TimerTask() {
+                    @Override
+                    public void run() {
+                        progressCallback.onUpdate((int) tempBody.getSize());
```
Comment (cketti) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java]: There's no guarantee the file size will reflect the actual number of bytes written before the `FileOutputStream` has been closed. I'd prefer to wrap the stream in a `CountingOutputStream` and get the count from there.

File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java [hunk 7]
```diff
@@ -990,24 +993,37 @@ public static boolean isDefaultMimeType(String mimeType) {
...
+                    }
+                }, 0, 50);
+            }
             IOUtils.copy(in, out);
         } finally {
+            if(timer != null){
```
Comment (cketti) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeUtility.java]: Code style: Missing spaces.

File: k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/FetchPartCallback.java [hunk 8]
```diff
@@ -28,7 +35,10 @@ public Object foundLiteral(ImapResponse response, FixedLengthInputStream literal
             String contentType = mPart
                     .getHeader(MimeHeader.HEADER_CONTENT_TYPE)[0];
 
-            return MimeUtility.createBody(literal, contentTransferEncoding, contentType);
+            if(attachmentProgressCallback != null)
```
Comment (cketti) [inline k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/FetchPartCallback.java]: Code style: Missing spaces and curly braces.

File: k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+import android.content.DialogInterface;
+import android.os.Bundle;
+
+import com.fsck.k9.controller.MessagingController;
+import com.fsck.k9.controller.MessagingListener;
+import com.fsck.k9.controller.SimpleMessagingListener;
+
+public class AttachmentDownloadDialogFragment extends DialogFragment {
+
+    static ProgressDialog dialog;
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java]: We absolutely don't want static fields in Fragments.

File: k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        args.putString(ARG_MESSAGE, message);
+        attachmentDownloadDialogFragment.setArguments(args);
+
+        return attachmentDownloadDialogFragment;
+    }
+
+    public void updateDownloadProgress(int newProgress) {
+        progress = newProgress;
+        for (int i = progress; i <= newProgress; i++) {
+            dialog.setProgress(i);
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java]: Calling `setProgress()` with intermediate values shouldn't be necessary.

File: k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        }
+    }
+
+    @Override
+    public Dialog onCreateDialog(Bundle savedInstanceState) {
+        Bundle args = getArguments();
+        int size = args.getInt(ARG_SIZE);
+        String message = args.getString(ARG_MESSAGE);
+
+        messagingListener = new SimpleMessagingListener(){
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java]: Code style: Missing space.

File: k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java [hunk 12]
```diff
@@ -165,7 +166,7 @@ public void onDestroy() {
 
     @Override
     public View onCreateView(LayoutInflater inflater, ViewGroup container,
-            Bundle savedInstanceState) {
+                             Bundle savedInstanceState) {
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java]: Wrong indentation. Import our code style settings to stop the IDE from making these changes.

File: k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java [hunk 13]
```diff
@@ -676,7 +678,7 @@ public void redisplayMessage() {
 
         @Override
         public void startPendingIntentForCryptoPresenter(IntentSender si, Integer requestCode, Intent fillIntent,
-                int flagsMask, int flagValues, int extraFlags) throws SendIntentException {
+                                                         int flagsMask, int flagValues, int extraFlags) throws SendIntentException {
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java]: Code style: Wrong indentation.

File: k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java [hunk 14]
```diff
@@ -781,7 +783,7 @@ public void run() {
 
         @Override
         public void startIntentSenderForMessageLoaderHelper(IntentSender si, int requestCode, Intent fillIntent,
-                int flagsMask, int flagValues, int extraFlags) {
+                                                            int flagsMask, int flagValues, int extraFlags) {
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java]: Code style: Wrong indentation.

File: k9mail-library/src/main/java/com/fsck/k9/mail/Folder.java [hunk 15]
```diff
@@ -129,7 +129,7 @@ public abstract void fetch(List<T> messages, FetchProfile fp,
                                MessageRetrievalListener<T> listener) throws MessagingException;
 
     public void fetchPart(Message message, Part part,
-                          MessageRetrievalListener<Message> listener) throws MessagingException {
+                          MessageRetrievalListener<Message> listener, AttachmentProgressCallback progressCallback) throws MessagingException {
```
Comment (cketti) [inline k9mail-library/src/main/java/com/fsck/k9/mail/Folder.java]: Code style: Wrong indentation.

File: k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/FetchPartCallback.java [hunk 16]
```diff
@@ -12,11 +13,17 @@
 
 class FetchPartCallback implements ImapResponseCallback {
     private Part mPart;
+    private AttachmentProgressCallback attachmentProgressCallback;
 
     FetchPartCallback(Part part) {
         mPart = part;
     }
 
+    FetchPartCallback(Part mPart, AttachmentProgressCallback attachmentProgressCallback) {
```
Comment (cketti) [inline k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/FetchPartCallback.java]: Code style: The parameter name shouldn't be prefixed with "m".

File: k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java (hunk 10 above)
Comment (harjot-oberai) [inline k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java]: I added this just to make the transition smoother. Its not actually needed, will remove it shortly.

File: k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java (hunk 12 above)
Comment (harjot-oberai) [inline k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java]: Well, the `settings.jar` given [here](https://github.com/k9mail/k-9/wiki/CodeStyle) broke the IDE 😅 . I am running Studio version 2.2.1. For the time being, I have manually corrected the indentation.

**Comment by harjot-oberai** [comment]
@cketti : Implemented all the requested changes 👍 .

File: k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+import com.fsck.k9.controller.MessagingController;
+import com.fsck.k9.controller.MessagingListener;
+import com.fsck.k9.controller.SimpleMessagingListener;
+
+public class AttachmentDownloadDialogFragment extends DialogFragment {
+
+    ProgressDialog dialog;
+
+    static int progress = 0;
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java]: This is still a `static` field.

Comment (harjot-oberai) [inline k9mail/src/main/java/com/fsck/k9/fragment/AttachmentDownloadDialogFragment.java]: Don't know how I skipped this 😅 . Fixed 👍

**Comment by cketti** [comment]
This crashes for me when pressing the "Save" button.

**Comment by AllanWang** [comment]
I would like to continue working on this pull request. Design wise, would K-9 like to keep it as a dialog or introduce a small circular loading bar next to the attachment so a user can still scroll and see the message while the attachment is downloading?

**Comment by harjot-oberai** [comment]
@cketti : I'm unable to reproduce the issue. Do you have the crash log ?

**Comment by cketti** [comment]
Stacktrace of the crash:

```
java.lang.NullPointerException: Attempt to read from field 'long com.fsck.k9.mailstore.AttachmentViewInfo.size' on a null object reference
   at com.fsck.k9.ui.messageview.MessageViewFragment.showDialog(MessageViewFragment.java:538)
   at com.fsck.k9.ui.messageview.MessageViewFragment.showAttachmentLoadingDialog(MessageViewFragment.java:656)
   at com.fsck.k9.ui.messageview.AttachmentController.downloadAttachment(AttachmentController.java:99)
   at com.fsck.k9.ui.messageview.AttachmentController.downloadAndSaveAttachmentTo(AttachmentController.java:85)
   at com.fsck.k9.ui.messageview.AttachmentController.saveAttachmentTo(AttachmentController.java:133)
   at com.fsck.k9.ui.messageview.AttachmentController.saveAttachmentTo(AttachmentController.java:72)
   at com.fsck.k9.ui.messageview.AttachmentController.saveAttachment(AttachmentController.java:68)
   at com.fsck.k9.ui.messageview.MessageViewFragment.onSaveAttachment(MessageViewFragment.java:809)
   at com.fsck.k9.ui.messageview.AttachmentView.onSaveButtonClick(AttachmentView.java:121)
   at com.fsck.k9.ui.messageview.AttachmentView.onClick(AttachmentView.java:100)
   at android.view.View.performClick(View.java:5609)
   at android.view.View$PerformClick.run(View.java:22262)
   at android.os.Handler.handleCallback(Handler.java:751)
   at android.os.Handler.dispatchMessage(Handler.java:95)
   at android.os.Looper.loop(Looper.java:154)
   at android.app.ActivityThread.main(ActivityThread.java:6077)
   at java.lang.reflect.Method.invoke(Native Method)
   at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

**Comment by cketti** [comment]
Continued as #2500

**Pull request closed without merging**